### PR TITLE
refactor: improve app startup (WPB-274)

### PR DESCRIPTION
**Pull request by ohassine**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-274" title="WPB-274" target="_blank">BugWPB-274</a>  Wire startup time
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Description

**The idea is to use Data Store instead of fetching database on app startup.**

Currently we access the database to check if we have a valid session and decide which screen to go. We are doing this on app startup, `onCreate()` and inside `runBlocking` block. This will block the thread causing many ANRs and affecting slow app startup (database init takes long time).

With this PR, I am replacing this by running `isLoggedInUser` use case that will check, using Data Store, if we have some users logged in and decide where to go. 
This will have a good impact on app startup as data store initialization is not that expensive. 
We will also get rid of many ANRs happening because of DB initialization.

The value stored in DataStore is updating constantly, as we are observing valid sessions in Wire Activity.

Before:

Screenshot 2023-09-26 at 17 49 06

 After

Screenshot 2023-09-26 at 17 33 45

### Dependencies (Optional)

Needs https://github.com/wireapp/kalium/pull/2047

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [968](https://10.10.124.134/job/AR-build-pipeline/968/) **failed**.

**Comment by AndroidBob** [comment]
Build [973](https://10.10.124.134/job/AR-build-pipeline/973/) **failed**.

**Comment by AndroidBob** [comment]
Build [979](https://10.10.124.134/job/AR-build-pipeline/979/) **failed**.

**Comment by AndroidBob** [comment]
Build [983](https://10.10.124.134/job/AR-build-pipeline/983/) **failed**.

**Comment by AndroidBob** [comment]
Build [992](https://10.10.124.134/job/AR-build-pipeline/992/) **failed**.

**Comment by AndroidBob** [comment]
Build [1002](https://10.10.124.134/job/AR-build-pipeline/1002/) **failed**.

**Review by alexandreferris — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [1006](https://10.10.124.134/job/AR-build-pipeline/1006/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.5.0(7677734)(7677734)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.5.0(7677734)(7677734)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt [hunk 1]
```diff
@@ -139,13 +148,23 @@ class WireActivityViewModel @Inject constructor(
         observeUpdateAppState()
         observeNewClientState()
         observeScreenshotCensoringConfigState()
+        observeAccounts()
+    }
+
+    private fun observeAccounts() {
+        viewModelScope.launch {
+            observeValidAccounts().distinctUntilChanged().collectLatest {
+                updateLoggedInUsersCount(it.size)
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt:166]: this looks like something that could be done in kalium side only. What's the point of having it in AR, user doesn't affect it anyway. 

We have some examples in `UserSessionScope` in the `init` blick we launch the coroutines that are supposed to run in background without users impact.

**Comment by AndroidBob** [comment]
Build [1012](https://10.10.124.134/job/AR-build-pipeline/1012/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.5.0(7683541)(7683541)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.5.0(7683541)(7683541)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/migration/MigrationManager.kt [hunk 2]
```diff
@@ -147,11 +154,17 @@ class MigrationManager @Inject constructor(
...
+                        userId.value,
+                        UserMigrationStatus.Successfully
+                    )
                     MigrationData.Result.Success
                 })
         } catch (e: Exception) {
-            globalDataStore.setUserMigrationStatus(userId.value, UserMigrationStatus.CompletedWithErrors)
+            globalDataStore.setUserMigrationStatus(
+                userId.value,
+                UserMigrationStatus.CompletedWithErrors
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/migration/MigrationManager.kt:165]: Why are all these code formatting ops happening?

File: app/src/test/kotlin/com/wire/android/ui/home/conversationslist/ConversationListViewModelTest.kt [hunk 3]
```diff
@@ -148,6 +149,7 @@ class ConversationListViewModelTest {
                 userTypeMapper = UserTypeMapper(),
                 updateConversationArchivedStatus = updateConversationArchivedStatus
             )
+        coEvery { observeConversationListDetailsUseCase(false) } returns flowOf(listOf())
```
Comment (gongracr) [inline app/src/test/kotlin/com/wire/android/ui/home/conversationslist/ConversationListViewModelTest.kt:171]: This one is overwriting the L130 above. Leftover?

**Comment by AndroidBob** [comment]
Build [1757](https://10.10.124.134/job/AR-build-pipeline/1757/) **failed**.

File: app/src/main/kotlin/com/wire/android/migration/MigrationManager.kt (hunk 2 above)
Comment (ohassine) [reply app/src/main/kotlin/com/wire/android/migration/MigrationManager.kt:165]: Auto reformat code, mostly reverted

**Comment by AndroidBob** [comment]
Build [1761](https://10.10.124.134/job/AR-build-pipeline/1761/) **failed**.

**Comment by AndroidBob** [comment]
Build [1763](https://10.10.124.134/job/AR-build-pipeline/1763/) **failed**.

**Comment by AndroidBob** [comment]
Build [1772](https://10.10.124.134/job/AR-build-pipeline/1772/) **failed**.

**Pull request closed without merging**